git_machete/cmd.py

```diff
@@ -339,359 +339,1027 @@ def popen_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs:
+                self.roots += [b]
 
-def back_up_definition_file() -> None:
-    shutil.copyfile(definition_file_path, definition_file_path + "~")
+        if not invalid_branches:
+            return
 
+        if len(invalid_branches) == 1:
+            ans: str = ask_if(self.cli_ctxt,
+                              f"Skipping `{invalid_branches[0]}` " +
+                              "which is not a local branch (perhaps it has been deleted?).\n" +
+                              "Slide it out from the definition file?" +
+                              pretty_choices("y", "e[dit]", "N"), opt_yes_msg=None)
+        else:
+            ans = ask_if(self.cli_ctxt,
+                         f"Skipping {', '.join(f'`{b}`' for b in invalid_branches)} " +
+                         "which are not local branches (perhaps they have been deleted?).\n" +
+                         "Slide them out from the definition file?" +
+                         pretty_choices("y", "e[dit]", "N"), opt_yes_msg=None)
+
+        def recursive_slide_out_invalid_branches(b: str) -> List[str]:
+            new_down_branches = flat_map(recursive_slide_out_invalid_branches, self.down_branches.get(b, []))
+            if b in invalid_branches:
+                if b in self.down_branches:
+                    del self.down_branches[b]
+                if b in self.annotations:
+                    del self.annotations[b]
+                if b in self.up_branch:
+                    for d in new_down_branches:
+                        self.up_branch[d] = self.up_branch[b]
+                    del self.up_branch[b]
+                else:
+                    for d in new_down_branches:
+                        del self.up_branch[d]
+                return new_down_branches
+            else:
+                self.down_branches[b] = new_down_branches
+                return [b]
 
-def save_definition_file() -> None:
-    with open(definition_file_path, "w") as f:
-        f.write("\n".join(render_tree()) + "\n")
+        self.roots = flat_map(recursive_slide_out_invalid_branches, self.roots)
+        self.managed_branches = excluding(self.managed_branches, invalid_branches)
+        if ans in ('y', 'yes'):
+            self.save_definition_file()
+        elif ans in ('e', 'edit'):
+            MacheteClient.edit(self.cli_ctxt)
+            self.read_definition_file(verify_branches)
+
+    def render_tree(self) -> List[str]:
+        if not self.indent:
+            self.indent = "\t"
+
+        def render_dfs(b: str, depth: int) -> List[str]:
+            self.annotation = f" {self.annotations[b]}" if b in self.annotations else ""
+            res: List[str] = [depth * self.indent + b + self.annotation]
+            for d in self.down_branches.get(b, []):
+                res += render_dfs(d, depth + 1)
+            return res
+
+        total: List[str] = []
+        for r in self.roots:
+            total += render_dfs(r, depth=0)
+        return total
+
+    @staticmethod
```

Comment:
If `definition_file_path` is converted to an instance attribute (which I think it should), then this should no longer be a static method